### ISSUE TITLE
[todo] Change `handle` functions to return enum instead of boolean

### ISSUE THREAD

**Issue by hellendag** [facebook/draft-js#339]
Currently, [`handle` functions](http://facebook.github.io/draft-js/docs/advanced-topics-key-bindings.html#customization) return a boolean to indicate to the Draft handler core whether the event has been handled or not. I wrote this code, and even I still forget whether `true` means that it was handled or deferred. 🙊 

To remedy this, I propose that `handle` functions instead return values of a simple binary enum, much the same way that `DraftRemovalDirection` clarifies the direction of a deletion. This should allow developers to know immediately which value to return, eliminating confusion about the outcome of the handler.

For instance, the enum could look like:

``` js
/* @flow */
export type DraftHandleValue = 'handled' | 'not-handled';
```

And a handler function could look like:

``` js
handleReturn(e: SyntheticKeyboardEvent): DraftHandleValue {
  // do stuff
  return 'handled';
}
```

This would be a breaking change, but definitely worthwhile.

This should be a pretty easy PR for anyone looking to get started with the codebase. It would require a new module to export the Flow type, plus changes at all the handler callsites to accommodate the non-boolean values.

Referenced at https://github.com/facebook/draft-js/issues/141#issuecomment-214408348.

**Comment by grrowl** [comment]
Surely it should mirror native DOM's behaviour, which is `return false` -> don't perform additional actions (or, `preventDefault`)?

**Comment by guitardave24** [comment]
Agree with Tom. I feel it's far too prominent in handler type code to
return undefined (not intentionally and not) to allow propagation. Having
to go through all handler code to return an explicit valid to represent not
handled seems fairly gratuitous.

**Comment by guitardave24** [comment]
To that end it might be worth allowing === true as well and not make it a
breaking change. (On phone haven't looked at PR so maybe you're already
doing this.)

**Comment by hellendag** [comment]
As I noted, I find boolean returns confusing, whether it mirrors native DOM behavior or not. Apart from legacy reasons, why is it a more correct convention to return `false` than `true` when no further actions should be performed? :) (It's also worth noting that as of React 14, return values in event handlers no longer have meaning [[#](https://facebook.github.io/react/docs/events.html)].)

As for returning `undefined`, the Flow type definitions for these handler functions currently insist upon boolean returns (https://github.com/facebook/draft-js/blob/master/src/component/base/DraftEditorProps.js#L88), though of course I can't guarantee that Draft users are using Flow. In theory, though, everyone should return explicit values.

My main point is that I am in favor of clarity and explicitness, even if it adds a little more typing. :)

**Comment by guitardave24** [comment]
I agree with all of your points and to the React 14 note, a more verbose (and perhaps correct) change would be to pass in a Draft-specific event where you can explicitly call a sort of `preventDefault()` on, and not rely on return values at all.

I suppose now is the time to make such a change, and accept that it will cause likely some migration pain to existing clients without a legacy fallback for `undefined`.

**Comment by hellendag** [comment]
> ...a more verbose (and perhaps correct) change would be to pass in a Draft-specific event where you can explicitly call a sort of preventDefault() on, and not rely on return values at all.

I've considered this, though I'm leery of creating new event-ish objects for it, with their own APIs to understand and document.

> I suppose now is the time to make such a change as you suggest, and accept that it will cause likely some migration pain to existing clients without a legacy fallback for undefined.

Yeah, I figure it's better to surface breaking changes now while the framework only been out for a couple months. :)

**Comment by guitardave24** [comment]
Agreed. Are you planning to work on this today @hellendag?

**Comment by pioul** [comment]
This issue can probably be closed: b4afd3c6aa5dd7203362beff3d1b020738a69cf6 (in since v0.9.0)

**Comment by mitermayer** [comment]
This issue has been fixed on https://github.com/facebook/draft-js/pull/340
